altinn-client/src/main/java/no/difi/meldingsutveksling/AltinnPackage.java

```diff
@@ -1,105 +1,203 @@
     }
 
     /**
      * Writes the Altinn package as a Zip file
      *
      * @param outputStream where the Zip file is written
-     * @throws IOException
      */
-    public void write(OutputStream outputStream) throws IOException {
+    public void write(OutputStream outputStream, ApplicationContext context) throws IOException {
         ZipOutputStream zipOutputStream = new ZipOutputStream(outputStream);
 
-        zipOutputStream.putNextEntry(new ZipEntry("manifest.xml"));
+        zipOutputStream.putNextEntry(new ZipEntry(MANIFEST_XML));
         marshallObject(manifest, zipOutputStream);
         zipOutputStream.closeEntry();
 
-        zipOutputStream.putNextEntry(new ZipEntry("recipients.xml"));
+        zipOutputStream.putNextEntry(new ZipEntry(RECIPIENTS_XML));
         marshallObject(recipient, zipOutputStream);
         zipOutputStream.closeEntry();
 
-        zipOutputStream.putNextEntry(new ZipEntry("content.xml"));
-        no.difi.meldingsutveksling.domain.sbdh.ObjectFactory objectFactory = new no.difi.meldingsutveksling.domain.sbdh.ObjectFactory();
-        marshallObject(objectFactory.createStandardBusinessDocument(eduDocument), zipOutputStream);
-        zipOutputStream.closeEntry();
+
+        if (sbd.getAny() instanceof Payload || sbd.getAny() instanceof JAXBElement) {
+            zipOutputStream.putNextEntry(new ZipEntry(CONTENT_XML));
+            no.difi.meldingsutveksling.domain.sbdh.ObjectFactory objectFactory = new no.difi.meldingsutveksling.domain.sbdh.ObjectFactory();
+            marshallObject(objectFactory.createStandardBusinessDocument(sbd), zipOutputStream);
+            zipOutputStream.closeEntry();
+
+            if (sbd.getAny() instanceof Payload) {
+                Payload payload = (Payload) sbd.getAny();
+                if (payload.getInputStream() != null) {
+                    zipOutputStream.putNextEntry(new ZipEntry(ASIC_FILE));
+                    IOUtils.copy(payload.getInputStream(), zipOutputStream);
+                    zipOutputStream.closeEntry();
+                }
+            }
+        }
+
+        if (sbd.getAny() instanceof BusinessMessage) {
+            zipOutputStream.putNextEntry(new ZipEntry(ALTINN_SBD_FILE));
+            ObjectMapper om = context.getBean(ObjectMapper.class);
+            om.configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false);
+            om.writeValue(zipOutputStream, sbd);
+            zipOutputStream.closeEntry();
+
+            if (this.asicInputStream != null) {
+                zipOutputStream.putNextEntry(new ZipEntry(ASIC_FILE));
+                IOUtils.copy(this.asicInputStream, zipOutputStream);
+                zipOutputStream.closeEntry();
+            }
+        }
 
         zipOutputStream.finish();
-        zipOutputStream.flush();
-        zipOutputStream.close();
+    }
+
+    public static AltinnPackage from(File f, MessagePersister messagePersister, ApplicationContext context) throws IOException, JAXBException {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to throw at most one checked exception instead of: java.io.IOException, javax.xml.bind.JAXBException [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-app01.difi.local:9000/coding_rules#rule_key=squid%3AS1160)